### ISSUE TITLE
Support general scripting or "execute series of commands"

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-spark-connector#159]
So far our current plan is going in the direction of sinking dataframes using this API.  That's good.  A thing that I've already found I'm going to need is a way of just executing an ordered series of cypher queries in a single partition (single-threaded).  Think of this as a way of executing a script.

Before we get into ways to do this -- 

# Use cases for why I want to do this:
* Schema setup / creating indexes prior to a load.  Daily loads might want to begin with deleting a subsection of yesterday's data, for example
* Executing a series of GDS actions (for example creating a virtual graph, and then getting results back from it)
* Cypher scripting from spark (i.e. storing & applying dynamic cypher in other data structures like tables)

This will enhance the ability to treat a single spark job as a big unit of work.  

Question is how best to do this.   Here's a sketch of how I've thought to do this.  Just make scripts dataframes and make the cypher runnable from a column, like so:

## Option 1 - Use "Query"

```
from pyspark.sql import Row
from pyspark.sql.functions import col

script_row = Row("sequence", "cypher", "param1")
my_script = [
    script_row(1, 'CREATE INDEX ON :Thing(id)', None),
    script_row(2, 'CREATE (:Thing { id: event.param1 })', 25)
]
​
script = spark.createDataFrame(data)

script.repartition(1).orderBy('sequence').write
  .format("org.neo4j.spark.DataSource")
  .option("url", "bolt://localhost:7687")
  .option("query", col('cypher'))
  .save()
```

Notice how the query is a DF column not a string.  The query format already gives us the ability to bind the entire row effectively as a set of parameters to the query.  This approach would be problematic with batching though (UNWIND events as event) and we don't want the user to have to remember to set batching=1 every time.   The other approach is a different explicit runmode, like this:

## Option 2 - Introduce "Script" (or something similar)
```
script.write
  .format("org.neo4j.spark.DataSource")
  .option("url", "bolt://localhost:7687")
  .option("script", True)
  .option("script.order", "sequence")
  .option("script.cypher", "cypher")
  .save()
```

This mode would *never* batch and would *always* repartition(1) and execute in the order of the `script.order` column, or else in the order of the first column.  Would always run in "write mode".  Would fail if script.cypher isn't specified.   But like "query", it would always bind the dataframe row to "event", so that by putting extra columns in the dataframe, a user could dynamically pass parameters to themselves.

## Alternatives

I already implemented a work-around approach for this - where I passed cypher in a dataframe as above, and then used `apoc.cypher.doIt()`.  This didn't work well for many reasons:
* Requires APOC dynamic cypher methods, which aren't available in all environments
* Due to security restrictions and changing security in neo4j procedures, apoc.cypher.doIt is going to have problems with certain privileged commands in certain environments (things like creating indexes, users, databases, etc)
* Requires the user to know too much about dynamic cypher, and c'mon...it's a work-around.  :) 

Thoughts?

**Comment by moxious** [comment]
(Note to myself to reach out to Mark N. about a collaborative GDS example when we can get this in)

**Comment by conker84** [comment]
After looking more deeply at this let me share some notes:

* we cannot `repartition` the Dataframe from the inside; it's an operation that the user must provide from his job;
* we cannot access to values inside the Dataframe `outside` the partition this means that we cannot leverage the `event` parameter if we want to execute it in the Spark Driver program (the `coordinator` of the cluster).

```python
script.repartition(1).orderBy('sequence').write \
  .format("org.neo4j.spark.DataSource") \
  .option("url", "bolt://localhost:7687") \
  .option("query", col('cypher')) \
  .save()
```

Pass a column to an option is not possible as it only takes string, number and boolean as value types.

We can do something like this
```python
script.write \
  .format("org.neo4j.spark.DataSource") \
  .option("url", "bolt://localhost:7687") \
  .option("script", True) \
  .option("script", "<MY_CYPHER_QUERY_WITHOUT_PARAMETERS>; <MY_OTHER_CYPHER_QUERY_WITHOUT_PARAMETERS>; ...") \
  .save()
```

This will be executed in the Driver program once before the import process starts but we can't share the return with the partitions because it should be consumed into the Driver with the risk of an OOM 
Do it fits with what you have in mind?

**Comment by moxious** [comment]
I like this approach: 

```
  .option("script", True) \
  .option("script", "<MY_CYPHER_QUERY_WITHOUT_PARAMETERS>; <MY_OTHER_CYPHER_QUERY_WITHOUT_PARAMETERS>; ...") \
  .save()
```

Parameters don't matter here, since if they need parameters, it's straightforward to use an existing dataframe to pipe those in as parameters.

The only change I'd make is to make the option an array of SINGLE statements, not one big text chunk.    

Would it be possible to return the results from the last command only, no fancy schema stuff, just as a value with a map of results?

**Comment by conker84** [comment]
> I like this approach:
> 
> ```
>   .option("script", True) \
>   .option("script", "<MY_CYPHER_QUERY_WITHOUT_PARAMETERS>; <MY_OTHER_CYPHER_QUERY_WITHOUT_PARAMETERS>; ...") \
>   .save()
> ```
> 
> Parameters don't matter here, since if they need parameters, it's straightforward to use an existing dataframe to pipe those in as parameters.
> 
> The only change I'd make is to make the option an array of SINGLE statements, not one big text chunk.
> 

Unfortunately is not possible as it only takes string, number and boolean as value types.

> Would it be possible to return the results from the last command only, no fancy schema stuff, just as a value with a map of results?

Yes but it will be a list of maps (one map per row returned by the query)

**Comment by moxious** [comment]
OK, then I guess we work with those limitations

**Comment by conker84** [comment]
closed via ee31233
